### PR TITLE
Target Flexible Data Models

### PR DESCRIPTION
**Pull request by einarmo**
Initial version of feature writing to flexible data models through DMSv3.

What this does is create a view and container for each referenced OPC-UA type, then map instances to those types, in the simplest possible terms.

It will likely need to evolve a great deal in the future. This is not made available to users yet.

### CODE REVIEW

File: Extractor/NodeSources/NodeSetSource.cs [hunk 1]
```diff
@@ -136,7 +142,7 @@ private void LoadReferences()
                     {
                         references[targetId] = targetRefs = new Dictionary<(NodeId, NodeId, bool), IReference>();
                     }
-                    if (!targetRefs.ContainsKey((node.NodeId, reference.ReferenceTypeId, reference.IsInverse)))
+                    if (!targetRefs.ContainsKey((node.NodeId, reference.ReferenceTypeId, !reference.IsInverse)))
```
Comment (ozangoktan) [inline Extractor/NodeSources/NodeSetSource.cs:145]: The flag here seems to have flipped, but I could not see an obvious change upstream of where it comes from. Raising it to confirm.

File: Extractor/NodeSources/NodeSetSource.cs [hunk 2]
```diff
@@ -362,7 +412,7 @@ public void BuildNodes(IEnumerable<NodeId> rootNodes, bool isFullBrowse)
                 isFullBrowse);
         }
 
-        private void GetRelationshipData()
+        private void GetRelationshipData(bool getPropertyReferences, HashSet<NodeId> additionalKnownNodes, List<(NodeId, IReference)> newNodes)
```
Comment (ozangoktan) [inline Extractor/NodeSources/NodeSetSource.cs:415]: Minor: Method name starts with get despite returning void, and no obvious out parameters

File: Extractor/NodeSources/NodeSetSource.cs [hunk 3]
```diff
@@ -330,9 +335,54 @@ public void BuildNodes(IEnumerable<NodeId> rootNodes, bool isFullBrowse)
                 InitNodeState(update, node);
             }
 
+        }
+
+        public override async Task<NodeSourceResult?> ParseResults(CancellationToken token)
+        {
+            if (!NodeMap.Any()) return null;
+
+            FinalDestinationObjects.Clear();
```
Comment (ozangoktan) [inline Extractor/NodeSources/NodeSetSource.cs:344]: Should we implement locks on these objects? They are referenced from a lot of places.

File: Extractor/Nodes/BaseUANode.cs [hunk 4]
```diff
@@ -152,6 +152,7 @@ public IEnumerable<BaseUANode> GetAllProperties()
 
         public BaseUANode(NodeId id, BaseUANode? parent, NodeId? parentId)
         {
+            // if (id.IsNullNodeId) throw new ArgumentException("Id may not be the null NodeId", nameof(id));
```
Comment (ozangoktan) [inline Extractor/Nodes/BaseUANode.cs:155]:
```suggestion
```

File: ConfigurationTool/Checks/Events.cs [hunk 5]
```diff
@@ -63,24 +63,37 @@ public async Task GetEventConfig(CancellationToken token)
...
-
-            var emittedEvents = referencedEvents.ToList();
+            var emittedEvents = emitterReferences.Select(evt => evt.Id).ToList();
 
             if (emittedEvents.Any())
             {
                 log.LogInformation("Identified {Count} events by looking at GeneratesEvent references", emittedEvents.Count);
+                foreach (var rf in emitterReferences)
+                {
+                    log.LogInformation("Test {T} {E}", rf.ParentId, rf.Id);
```
Comment (ozangoktan) [inline ConfigurationTool/Checks/Events.cs:95]: Maybe this should be moved to a more verbose log level?

File: Extractor/Pushers/CDFPusher.cs [hunk 6]
```diff
@@ -301,7 +319,14 @@ public sealed class CDFPusher : IPusher
 
             log.LogInformation("Testing {TotalNodesToTest} nodes against CDF", variables.Count() + objects.Count());
 
-            if (fullConfig.DryRun) return result;
+            if (fullConfig.DryRun)
+            {
+                if (fdmDestination != null)
+                {
+                    await fdmDestination.PushNodes(objects, variables, references, Extractor, token);
```
[the remark this answers is not in the document]
Comment (einarmo) [reply Extractor/Pushers/CDFPusher.cs:326]: Yeah, it's _very_ nice to get a dump of all the types and such when doing a dry run. Without this the dry run essentially doesn't show the user what we will make in FDM.

File: Extractor/Types/UAPropertySerializer.cs [hunk 7]
```diff
@@ -31,6 +31,20 @@
 
 namespace Cognite.OpcUa.Types
 {
+    public enum StringConverterMode
+    {
+        Simple,
+        Json,
+        ReversibleJson
```
[the remark this answers is not in the document]
Comment (einarmo) [reply Extractor/Types/UAPropertySerializer.cs:38]: So for serializing most OPC-UA types we use the JSON encoder from the OPC-UA SDK, which is somewhat finnicky, it wasn't really intended to be used this way. Reversible JSON is telling it to include type information in the payload, which is handy if we're going to convert it back to OPC-UA later.

File: Extractor/NodeSources/NodeSetSource.cs (hunk 1 above)
Comment (einarmo) [reply Extractor/NodeSources/NodeSetSource.cs:145]: It was a bug, actually. This feature never worked as it should... It is pretty niche.

File: Extractor/NodeSources/NodeSetSource.cs (hunk 3 above)
Comment (einarmo) [reply Extractor/NodeSources/NodeSetSource.cs:344]: They are protected, and I don't see any parallelism in this class, it shouldn't be necessary?

File: ConfigurationTool/Checks/Events.cs (hunk 5 above)
Comment (einarmo) [reply ConfigurationTool/Checks/Events.cs:95]: Should probably just be removed...

File: Extractor/NodeSources/NodeSetSource.cs (hunk 3 above)
Comment (ozangoktan) [reply Extractor/NodeSources/NodeSetSource.cs:344]: Should be fine then 👍🏼

File: Extractor/Pushers/CDFPusher.cs (hunk 6 above)
[the remark this answers is not in the document]
Comment (ozangoktan) [reply Extractor/Pushers/CDFPusher.cs:326]: Though I see the use case, would argue that it invalidates dry run. Maybe a config option could be put in place? 

**_--dry-run-with-type-schemas_** or something similar?

[the remark this answers is not in the document]
Comment (einarmo) [reply Extractor/Pushers/CDFPusher.cs:326]: Oh. It doesn't actually push, it's still a dry run.

[the remark this answers is not in the document]
Comment (ozangoktan) [reply Extractor/Pushers/CDFPusher.cs:326]: Ah, it does that internally. 🚀

File: Extractor/Types/UAPropertySerializer.cs (hunk 7 above)
[the remark this answers is not in the document]
Comment (einarmo) [reply Extractor/Types/UAPropertySerializer.cs:38]: This produces a very different type of JSON, I think having it abstracted from the outside is fine. The values produced by this contains some type information in certain cases.

[the remark this answers is not in the document]
Comment (ozangoktan) [reply Extractor/Types/UAPropertySerializer.cs:38]: Ah, looks like this is an OPCUA thing -> https://reference.opcfoundation.org/Core/Part6/v105/docs/5.4

👍🏼

**Review by ozangoktan — APPROVED** [review]

**Pull request merged**